### PR TITLE
Transactional Email API

### PR DESCRIPTION
**Pull request by hownowstephen**
Adds client support for the upcoming transactional email API

### CODE REVIEW

File: transactional.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	return e.Err
+}
+
+func (c *TransactionalAPIClient) SendEmail(e Email) (*TransactionalResponse, error) {
+	b, err := json.Marshal(e)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("https://%s/v1/send/email", c.Host), bytes.NewBuffer(b))
```
Comment (newhook) [inline transactional.go:46]: what if I want to use http?

File: transactional.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.Key))
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.Client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (newhook) [inline transactional.go:59]: close the body

File: email.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	Body                   string                 `json:"body,omitempty"`
+	MessageData            map[string]interface{} `json:"message_data,omitempty"`
+	BCC                    string                 `json:"bcc,omitempty"`
+	ReplyTo                string                 `json:"reply_to,omitempty"`
+	FromID                 int                    `json:"from_id,omitempty"`
+	ReplyToID              int                    `json:"reply_to_id,omitempty"`
+	Headers                map[string]string      `json:"headers,omitempty"`
+	Attachments            map[string]string      `json:"attachments,omitempty"`
+	FakeBCC                *bool                  `json:"fake_bcc"`
+	HideBody               *bool                  `json:"hide_body"`
```
Comment (idawes) [inline email.go:25]: I thought we'd decided on something other than HideBody that reflected more that we're not going to store the body.

File: transactional.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+type TransactionalResponse struct {
+	Recipient  string `json:"recipient"`
+	DeliveryID string `json:"delivery_id"`
+	QueuedAt   int    `json:"queued_at"`
```
Comment (newhook) [inline transactional.go:14]: queuedat should be `unix.Time`

File: examples/transactional.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+func main() {
+
+	client := customerio.NewAPIClient("<your-key-here>")
+
+	email := customerio.Email{
+		CustomerID: "customer_1",
+		To:         "customer@example.com",
+		From:       "business@example.com",
+		Subject:    "hello world",
+		Body:       "hello from the Customer.io Go Client",
```
Comment (newhook) [inline examples/transactional.go:18]: add an example with variables.

File: customerio.go [hunk 6]
```diff
@@ -32,6 +32,10 @@ func (e *CustomerIOError) Error() string {
 	return fmt.Sprintf("%v: %v %v", e.status, e.url, string(e.body))
 }
 
+func NewTrackClient(siteID, apiKey string) *CustomerIO {
```
Comment (newhook) [inline customerio.go:35]: Need a comment.

File: email.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
+package customerio
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"io"
+)
+
+type Email struct {
```
Comment (newhook) [inline email.go:10]: godoc comments.

File: README.md [hunk 8]
```diff
@@ -1,12 +1,12 @@
 # Customerio
+
```
Comment (newhook) [inline README.md:2]: Add docs for transactional (CHAD)

File: transactional.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+type TransactionalError struct {
+	Err    string
+	Status int
+}
+
+func (e *TransactionalError) Error() string {
+	return e.Err
+}
+
+func (c *APIClient) SendEmail(e Email) (*TransactionalResponse, error) {
```
Comment (newhook) [inline transactional.go:26]:
```
func (c* APIClient) SendEmail(ctx context.Context, request *SendEmailRequest) (*SendEmailResponse, error)
```

File: README.md (hunk 8 above)
Comment (hownowstephen) [reply README.md:2]: pinged him about it 👍

File: api.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func (c *APIClient) doRequest(ctx context.Context, verb, requestPath string, body interface{}) (*http.Response, []byte, error) {
+
+	b, err := json.Marshal(body)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	req, err := http.NewRequest("POST", c.URL+requestPath, bytes.NewBuffer(b))
```
Comment (newhook) [inline api.go:34]: This should use https://golang.org/pkg/net/http/#NewRequestWithContext

File: api.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+	req.Header.Set("Authorization", "Bearer "+c.Key)
+	req.Header.Set("Content-Type", "application/json")
+
+	var resp *http.Response
+	var responseBody []byte
+
+	errc := make(chan error, 1)
+	go func() {
+		resp, err = c.Client.Do(req)
```
Comment (newhook) [inline api.go:49]: This isn't necessary. See the above.

File: transactional.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	"time"
+)
+
+type TransactionalResponse struct {
+	Recipient  string    `json:"recipient"`
+	DeliveryID string    `json:"delivery_id"`
+	QueuedAt   time.Time `json:"queued_at"`
+}
+
+func (t *TransactionalResponse) UnmarshalJSON(b []byte) error {
```
Comment (newhook) [inline transactional.go:14]: ugh, is there no better way to do that?

File: README.md [hunk 13]
```diff
@@ -53,50 +53,50 @@ Create an instance of the client with your [customer.io](http://customer.io) cre
...
+  if you pass along the current subscription plan (free / basic / premium) for your customers, you can
+  set up triggers which are only sent to customers who have subscribed to a
+  particular plan (e.g. "premium").
 
-You'll want to indentify your customers when they sign up for your app and any time their
+You'll want to identify your customers when they sign up for your app and any time their
 key information changes. This keeps [Customer.io](http://customer.io) up to date with your customer information.
 
-```go
+````go
```
Comment (idawes) [inline README.md:80]: Why did we add a 4th ` here and in a bunch of other places?

File: README.md [hunk 14]
```diff
@@ -180,28 +178,52 @@ Deleting a device will remove it from the customers device list in Customer.io.
...
+  MessageData: map[string]interface{}{
+    "name": "",
+    "items": map[string]interface{}{
+      "name": "",
+      "price": "",
+    },
+    "products": []interface{}{},
+  },
+  Identifiers: map[string]interface{}{
+    "id": "",
```
Comment (idawes) [inline README.md:215]: this should have an example id instead of being blank

File: api.go (hunk 10 above)
Comment (idawes) [reply api.go:33]: are we planning on fixing this before shipping?

File: customerio_test.go [hunk 15]
```diff
@@ -1,155 +1,298 @@
...
+		panic(err)
 	}
 
-	os.Exit(exitCode)
-}
-
-func TestIdentify(t *testing.T) {
+	cio = customerio.NewCustomerIO("siteid", "apikey")
+	cio.Host = u.Host
+	cio.SSL = false
```
Comment (idawes) [inline customerio_test.go:32]: might be a good idea to comment this to make it clear that this is not desirable for production usage.

File: customerio_test.go [hunk 16]
```diff
@@ -1,155 +1,298 @@
...
-	if err != nil {
-		t.Error(err.Error())
+func TestIdentify(t *testing.T) {
+	attributes := map[string]interface{}{
+		"a": "1",
+	}
+	if err := cio.Identify("", attributes); err == nil {
+		t.Error("expected error")
+	} else {
+		checkParamError(t, err, "customerID")
```
Comment (idawes) [inline customerio_test.go:71]: checkParamError could encapsulate the nil check.

**Pull request merged**